// node-pg-migrate/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PgType;
(function (PgType) {
    PgType["BIGINT"] = "bigint";
    PgType["INT8"] = "int8";
    PgType["BIGSERIAL"] = "bigserial";
    PgType["BIT_1"] = "bit";
    PgType["BIT_VARYING"] = "bit varying";
    PgType["VARBIT"] = "varbit";
    PgType["SERIAL8"] = "serial8";
    PgType["BOOLEAN"] = "boolean";
    PgType["BOOL"] = "bool";
    PgType["BOX"] = "box";
    PgType["BYTEA"] = "bytea";
    PgType["CHARACTER"] = "character";
    PgType["CHAR"] = "char";
    PgType["CHARACTER_VARYING"] = "character varying";
    PgType["VARCHAR"] = "varchar";
    PgType["CIDR"] = "cidr";
    PgType["CIRCLE"] = "circle";
    PgType["DATE"] = "date";
    PgType["DOUBLE_PRECISION"] = "double precision";
    PgType["INET"] = "inet";
    PgType["INTEGER"] = "integer";
    PgType["INT"] = "int";
    PgType["INT4"] = "int4";
    PgType["INTERVAL"] = "interval";
    PgType["JSON"] = "json";
    PgType["JSONB"] = "jsonb";
    PgType["LINE"] = "line";
    PgType["LSEG"] = "lseg";
    PgType["MACADDR"] = "macaddr";
    PgType["MONEY"] = "money";
    PgType["NUMERIC"] = "numeric";
    PgType["PATH"] = "path";
    PgType["PG_LSN"] = "pg_lsn";
    PgType["POINT"] = "point";
    PgType["POLYGON"] = "polygon";
    PgType["REAL"] = "real";
    PgType["FLOAT4"] = "float4";
    PgType["SMALLINT"] = "smallint";
    PgType["INT2"] = "int2";
    PgType["SMALLSERIAL"] = "smallserial";
    PgType["SERIAL2"] = "serial2";
    PgType["SERIAL"] = "serial";
    PgType["SERIAL4"] = "serial4";
    PgType["TEXT"] = "text";
    PgType["TIME"] = "time";
    PgType["TIME_WITHOUT_TIME_ZONE"] = "without time zone";
    PgType["TIME_WITH_TIME_ZONE"] = "time with time zone";
    PgType["TIMETZ"] = "timetz";
    PgType["TIMESTAMP"] = "timestamp";
    PgType["TIMESTAMP_WITHOUT_TIME_ZONE"] = "timestamp without time zone";
    PgType["TIMESTAMP_WITH_TIME_ZONE"] = "timestamp with time zone";
    PgType["TIMESTAMPTZ"] = "timestamptz";
    PgType["TSQUERY"] = "tsquery";
    PgType["TSVECTOR"] = "tsvector";
    PgType["TXID_SNAPSHOT"] = "txid_snapshot";
    PgType["UUID"] = "uuid";
    PgType["XML"] = "xml";
})(PgType = exports.PgType || (exports.PgType = {}));
